### PR TITLE
Remove the use of dmlc::JSONWriter, dmlc::Stream, and dmlc serializer

### PR DESCRIPTION
**Pull request by hcho3**
Extracted from #285.

* Do not use `dmlc/json.h`, `dmlc/io.h`, and `dmlc/serializer.h`.
* Replace all uses of `dmlc::Stream` with `std::fstream` and `std::stringstream`.
* Replace `dmlc::JSONWriter` with `rapidjson::Writer`.
* Use `uint64_t` for storing instance counts in JSON.
* Rename `TaskParameter` to `TaskParam`, to be consistent with `ModelParam`.
* Replace `ReferenceSerialize` with `SerializeToJSON`. Previously, `ReferenceSerialize` was a reference implementation of serializer for tree models. `ReferenceSerialize` was useful for verifying the correctness of other serializer methods of the tree objects. The reference serializer had to go, unfortunately, because it used dmlc serializer functions. As a replacement, I implemented a JSON serializer `SerializeToJSON`. While not very efficient compared to binary serializers, the JSON serializer will be useful for debugging and testing.
* Reimplement `LoadText` with `std::istream` and `std::getline`.

### CODE REVIEW

File: include/treelite/annotator.h [hunk 1]
```diff
@@ -27,14 +29,14 @@ class BranchAnnotator {
   void Annotate(const Model& model, const DMatrix* dmat, int nthread, int verbose);
   /*!
    * \brief load branch annotation from a JSON file
-   * \param fi input stream
+   * \param fp input stream
    */
-  void Load(dmlc::Stream* fi);
+  void Load(FILE* fp);
```
Comment (trivialfis) [inline include/treelite/annotator.h]: Why not c++ stream?

File: src/frontend/xgboost.cc [hunk 2]
```diff
@@ -44,12 +43,21 @@ namespace {
 
 typedef float bst_float;
 
+template <typename T>
+inline T* BeginPtr(std::vector<T>& vec) {  // NOLINT(*)
+  if (vec.empty()) {
```
Comment (trivialfis) [inline src/frontend/xgboost.cc]: Can you use the data member function?

**Review by trivialfis — COMMENTED** [review]
Looked briefly.  Not a complete review.

File: include/treelite/annotator.h (hunk 1 above)
Comment (hcho3) [reply include/treelite/annotator.h]: Good point. Let me use `std::istream&` here, to be consistent with the changes made in `frontend/xgboost.cc` and `frontend/lightgbm.cc`.

File: src/frontend/xgboost.cc (hunk 2 above)
Comment (hcho3) [reply src/frontend/xgboost.cc]: Done. All uses of `BeginPtr` in `frontend/xgboost.cc` involve non-empty containers, so we can just use `data()` instead.

**Review by trivialfis — APPROVED** [review]

**Pull request merged**